handlers/product-switch-api/src/contributionToSupporterPlus.ts

```diff
@@ -0,0 +1,287 @@
+import { checkDefined } from '@modules/nullAndUndefined';
+import { zuoraDateFormat } from '@modules/zuora/common';
+import type { ZuoraClient } from '@modules/zuora/zuoraClient';
+import type { Dayjs } from 'dayjs';
+import dayjs from 'dayjs';
+import { removePendingUpdateAmendments } from './amendments';
+import type { CatalogInformation } from './catalogInformation';
+import { takePaymentOrAdjustInvoice } from './payment';
+import { sendThankYouEmail } from './productSwitchEmail';
+import { sendSalesforceTracking } from './salesforceTracking';
+import type {
+	ChangePlanOrderAction,
+	CreateOrderRequest,
+	OrderAction,
+	PreviewOrderRequest,
+	ZuoraPreviewResponse,
+	ZuoraSwitchResponse,
+} from './schemas';
+import {
+	zuoraPreviewResponseSchema,
+	zuoraSwitchResponseSchema,
+} from './schemas';
+import { sendToSupporterProductData } from './supporterProductData';
+import type { SwitchInformation } from './switchInformation';
+
+export type PreviewResponse = {
+	amountPayableToday: number;
+	contributionRefundAmount: number;
+	supporterPlusPurchaseAmount: number;
+	nextPaymentDate: string;
+};
+
+export const switchToSupporterPlus = async (
+	zuoraClient: ZuoraClient,
+	productSwitchInformation: SwitchInformation,
+): Promise<void> => {
+	const switchResponse = await doSwitch(zuoraClient, productSwitchInformation);
+
+	const paidAmount = await takePaymentOrAdjustInvoice(
+		zuoraClient,
+		switchResponse,
+		productSwitchInformation.catalog.supporterPlus.subscriptionChargeId,
+		productSwitchInformation.account.id,
+		productSwitchInformation.account.defaultPaymentMethodId,
+	);
+
+	await sendThankYouEmail(paidAmount, productSwitchInformation);
+
+	await sendSalesforceTracking(paidAmount, productSwitchInformation);
+
+	await sendToSupporterProductData(productSwitchInformation);
```

Comment:
No big deal either way, but do these need to happen in serial? Could this be:

```suggestion
	await Promise.all([
                sendThankYouEmail(paidAmount, productSwitchInformation),
                sendSalesforceTracking(paidAmount, productSwitchInformation),
                sendToSupporterProductData(productSwitchInformation),
         ]);
```

Reply:
Good point 👍